Mathlib/Probability/Kernel/IonescuTulcea/Traj.lean

```diff
@@ -0,0 +1,717 @@
+    IsProjectiveLimit ν μ ↔ ∀ ⦃n⦄, n ≥ a → ν.map (frestrictLe n) = μ (Iic n) := by
+  refine ⟨fun h n _ ↦ h (Iic n), fun h I ↦ ?_⟩
+  have := (I.subset_Iic_sup_id.trans (Iic_subset_Iic.2 (le_max_left (I.sup id) a)))
+  rw [← restrict₂_comp_restrict this, ← Measure.map_map, ← frestrictLe, h (le_max_right _ _), ← hμ]
+  all_goals fun_prop
+
+/-- To check that a measure `ν` is the projective limit of a projective family of measures indexed
+by `Finset ℕ`, it is enough to check on intervals of the form `Iic n`. -/
+theorem isProjectiveLimit_nat_iff {μ : (I : Finset ℕ) → Measure (Π i : I, X i)}
+    (hμ : IsProjectiveMeasureFamily μ) (ν : Measure (Π n, X n)) :
+    IsProjectiveLimit ν μ ↔ ∀ n, ν.map (frestrictLe n) = μ (Iic n) := by
+  rw [isProjectiveLimit_nat_iff' hμ _ 0]
+  simp
+
+variable (μ : (n : ℕ) → Measure (Π i : Iic n, X i))
+
+/-- Given a family of measures `μ : (n : ℕ) → Measure (Π i : Iic n, X i)`, we can define a family
+of measures indexed by `Finset ℕ` by projecting the measures. -/
+noncomputable def inducedFamily (S : Finset ℕ) : Measure ((k : S) → X k) :=
+    (μ (S.sup id)).map (restrict₂ S.subset_Iic_sup_id)
+
+instance [∀ n, SFinite (μ n)] (I : Finset ℕ) :
+    SFinite (inducedFamily μ I) := by rw [inducedFamily]; infer_instance
+
+instance [∀ n, IsFiniteMeasure (μ n)] (I : Finset ℕ) :
+    IsFiniteMeasure (inducedFamily μ I) := by rw [inducedFamily]; infer_instance
+
+instance [∀ n, IsZeroOrProbabilityMeasure (μ n)] (I : Finset ℕ) :
+    IsZeroOrProbabilityMeasure (inducedFamily μ I) := by rw [inducedFamily]; infer_instance
+
+instance [∀ n, IsProbabilityMeasure (μ n)] (I : Finset ℕ) :
+    IsProbabilityMeasure (inducedFamily μ I) := by
+  rw [inducedFamily]
+  exact isProbabilityMeasure_map (measurable_restrict₂ _).aemeasurable
+
+/-- Given a family of measures `μ : (n : ℕ) → Measure (Π i : Iic n, X i)`, the induced family
+equals `μ` over the intervals `Iic n`. -/
+theorem inducedFamily_Iic (n : ℕ) : inducedFamily μ (Iic n) = μ n := by
+  rw [inducedFamily, ← measure_cast (sup_Iic n) μ]
+  congr with x i
+  rw [restrict₂, cast_pi (by rw [sup_Iic n])]
+
+/-- Given a family of measures `μ : (n : ℕ) → Measure (Π i : Iic n, X i)`, the induced family
+will be projective only if `μ` is projective, in the sense that if `a ≤ b`, then projecting
+`μ b` gives `μ a`. -/
+theorem isProjectiveMeasureFamily_inducedFamily
+    (h : ∀ a b : ℕ, ∀ hab : a ≤ b, (μ b).map (frestrictLe₂ hab) = μ a) :
+    IsProjectiveMeasureFamily (inducedFamily μ) := by
+  intro I J hJI
+  have sls : J.sup id ≤ I.sup id := sup_mono hJI
+  simp only [inducedFamily]
+  rw [Measure.map_map, restrict₂_comp_restrict₂,
+    ← restrict₂_comp_restrict₂ J.subset_Iic_sup_id (Iic_subset_Iic.2 sls), ← Measure.map_map,
+    ← frestrictLe₂, h (J.sup id) (I.sup id) sls]
+  all_goals fun_prop
+
+end MeasureTheory
+
+end ProjectiveFamily
+
+variable (κ : (n : ℕ) → Kernel (Π i : Iic n, X i) (X (n + 1))) [∀ n, IsMarkovKernel (κ n)]
```

Comment:
Here you are declaring an explicit `κ` for the whole file. I see many applications of your lemmas in which you use `_` for that argument. That make me think that you should rather turn it into an implicit argument and switch to explicit only where it makes sense.